gk/fib.c

```diff
@@ -1664,3 +1675,445 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+static int
+list_ipv4_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	char *reply_msg, int *msg_len)
+{
+	int ret, index;
+
+	size_t lua_msg_len;
+	const char *lua_msg;
+
+	struct gk_fib *fib;
+	const struct rte_lpm_rule *re4;
+	struct rte_lpm_iterator_state state;
+	struct gk_fib_dump_entry dentry;
+
+	ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		luaL_error(l, "gk: failed to initialize the lpm rule iterator state at %s!",
+			__func__);
+		goto out;
+	}
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv4;
+		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
+		dentry.prefix_len = state.depth;
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0) {
+			luaL_error(l, "gk: failed to fillup the IPv4 GK FIB dump entry!");
+			goto out;
+		}
+
+		lua_pushvalue(l, 1);
+		lua_pushlightuserdata(l, &dentry);
+
+		if (lua_pcall(l, 1, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		lua_msg = luaL_checklstring(l, -1, &lua_msg_len);
+		if (lua_msg_len + *msg_len >= MSG_MAX_LEN) {
+			strncpy(reply_msg + *msg_len, lua_msg,
+				MSG_MAX_LEN - *msg_len);
+			reply_msg[MSG_MAX_LEN - 1] = '\0';
+			*msg_len = MSG_MAX_LEN;
+		} else {
+			strncpy(reply_msg + *msg_len, lua_msg, lua_msg_len);
+			*msg_len += lua_msg_len;
+		}
+
+		lua_pop(l, 1);
+
+		rte_free(dentry.ether_tbl);
+
+		index = rte_lpm_rule_iterate(&state, &re4);
+	}
+
+	lua_settop(l, 0);
+
+out:
+	return ret;
+}
+
+static int
+list_ipv6_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	char *reply_msg, int *msg_len)
+{
+	int ret, index;
+
+	size_t lua_msg_len;
+	const char *lua_msg;
+
+	struct gk_fib *fib;
+	struct rte_lpm6_rule re6;
+	struct rte_lpm6_iterator_state state6;
+	struct gk_fib_dump_entry dentry;
+
+	ret = rte_lpm6_iterator_state_init(ltbl->lpm6, NULL, 0, &state6);
+	if (ret < 0) {
+		luaL_error(l, "gk: failed to initialize the lpm6 rule iterator state at %s!",
+			__func__);
+		goto out;
+	}
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&dentry.addr.ip.v6, re6.ip,
+			sizeof(dentry.addr.ip.v6));
+		dentry.prefix_len = re6.depth;
+		fib = &ltbl->fib_tbl6[re6.next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0) {
+			luaL_error(l, "gk: failed to fillup the IPv6 GK FIB dump entry!");
+			goto out;
+		}
+
+		lua_pushvalue(l, 1);
+		lua_pushlightuserdata(l, &dentry);
+
+		if (lua_pcall(l, 1, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		lua_msg = luaL_checklstring(l, -1, &lua_msg_len);
+		if (lua_msg_len + *msg_len >= MSG_MAX_LEN) {
+			strncpy(reply_msg + *msg_len, lua_msg,
```

Comment:
Forcing the Lua callback function to return a string is complicating the C code **and** restricting what Lua scripts can do with `list_ipv4_fib_entries()` and `list_ipv6_fib_entries()`. The trick to clean things up is
1. to add a parameter `acc` (i.e. accumulator) in the callback function. The callback function does whatever it wants to `acc` and returns whatever it wants.
2. `list_ipv4_fib_entries()` and `list_ipv6_fib_entries()` receive one more parameter: `acc_start` that is passed to the first call of the callback function. All the falling calls of the callback function receives whatever the previous callback function returned.
3. `list_ipv4_fib_entries()` and `list_ipv6_fib_entries()` returns whatever that the last callback function returned.

In this new format, all the string managed is pushed to Lua and, in fact, returning a string becomes just a special case. For example, the callback function could compute the number of stale Ethernet entries.